### DISCUSSION TITLE
Differences between oxlint and biome?

### DISCUSSION THREAD

**Discussion by mcmxcdev** [oxc-project/oxc#1709] [Q&A] [answered]
I am following along the progress of oxlint and rome/biome for a while now and would like to understand why both tools, which are so similiar in [philosopy](https://github.com/oxc-project/oxc#-philosophy) are being developed as separate standalone tools.

According to https://npmtrends.com/@biomejs/biome-vs-oxlint, biome has 91k downloads per week while oxlint has 25k downloads per week, both tools are built in Rust, reimplement various ESLint rules and are probably encountering similar bugs and problems too.

I think it would be great to combine biome and oxlint into one, leading to higher velocity, more developer activity and less ecosystem fragmentation where devs have to get informed about both and pick one over the other for reason x or y.

**Comment by Boshen** [comment]
Give both a try and see which one fits your requirements. Biomes excel in every aspect at the moment.

**Reply by alexleduc76** [reply]
Biome is an opinionated formatter, which means it won't offer many formatting rules like ESLint does. I haven't tried OxLint yet, but if the features page is to be believed, it offers:

- Over 200 rules with a growing list from eslint, typescript, eslint-plugin-react, eslint-plugin-jest, eslint-plugin-unicorn and eslint-plugin-jsx-a11y.

As someone who cares a lot about picking my own formatting rules, OxLint looks like the Rust based linter I have been looking for.

Food for thought:
https://dev.to/alexleduc76/opinionated-code-formatters-suck-3b1

**Reply by mcmxcdev** [reply]
We are already using Biome (Rome before that) as part of our lint-staged setup (works great) and oxlint doesn't find any additional issues that were not discovered by Biome before.

My intention for this discussion was not to ask which tool to pick over the other (I already have) but to give food for thought to potentially merge or find out about the reasoning for not joining forces.

**Reply by alexleduc76** [reply]
I also did not mean my reply as a discussion on which to pick. I just wanted to highlight that this difference (opinionated VS flexible formatting rules)  is a perfectly valid reason for the projects to remain separate.

**Reply by Conaclos** [reply]
> As someone who cares a lot about picking my own formatting rules, OxLint looks like the Rust based linter I have been looking for.

In my knowledge OXC is working on a formatter compatible with Prettier. I am not sure that OXC will implement the now deprecated formatting rules of ESLint. Neither Biome, nor OXC satisfy your needs...

**Reply by alexleduc76** [reply]
The ESlint formatting rules have just been moved to an [ESlint plugin](https://eslint.style/packages/default), not deprecated. If OXC doesn't plan to implement those, then I guess I'll just keep using ESLint until someone makes a Rust port of it or something.

Edit: Never mind, you are right about the word "deprecated". Still, I'm glad the functionality is still there, even if it's spun off as a plug-in.
 https://eslint.org/blog/2023/10/deprecating-formatting-rules/

**Reply by alexleduc76** [reply]
Maybe the tool I'm looking for doesn't exist. Something that would solve this would be a "meta formatter" editor plug-in, that applies its own in-RAM-only representation of the code and on-the-fly converts the code in the format you prefer in your editor, but when you save the file, it separately runs the code through an opinionated formatter and that is what is saved to disk. This way, stylistic formatting rules would not affect Git commits(*), and people who hate the imposed opinionated formatting would have their cake and eat it.  *(talking about the annoying thing where someone overtake the git blame on every line of code changed by having different formatting rules than everyone else).

**Reply by nicod-pc** [reply]
@alexleduc76 I think you missed the advantage of prettier. In your article you bring the case whether `else` is in a separate line or not. For readability it doesn't really matter. Your argument is, that you are used to your way, so you want to keep it, which is fair. But if someone new joins you team they are maybe irritated by it. And as it's an option, you could format it differently. So you have a discussion with them. I spend hours in the past with colleagues discussion about how we want to format the code. Time that is wasted. What prettier achieved is to stop discussions. The whole world kind of just uses prettier now. So people don't choose prettier because they agree with their formatting, but because they don't have to think nor discuss about the formatting rules.

**Reply by alexleduc76** [reply]
@iwt-nduesing I would just shut down the discussion if anyone tried to challenge the formatting rules. The lead developer should make these decisions, and they should not be up for debate.

I have not missed the argument that Prettier's imposed formatting stops such discussions, but I disagree that it does. Someone could disagree just as much with the Prettier imposed formatting than the one I impose with ESLint rules, and argue that we should use another tool than Prettier for that reason. Prettier is not what shuts down formatting debates, it's the responsibility of the person with the authority to make those decisions.

**Reply by nicod-pc** [reply]
@alexleduc76 well, not everyone works in a top down hierarchy organization. This is the first time for years I see someone arguing for using a different tool than prettier for formatting, because they don't like their rules.

**Reply by Zwyx** [reply]
@alexleduc76, I know someone who is very much against Prettier, for the reasons you state. This is not well seen by many others, especially new devs joining their team.

In your article, you mention _Who are the opinionated formatter's authors to decide what's best for everyone else?_. I believe this is not how Prettier was made: Prettier's default config is the result of **hundreds of developers debating in [thousands of GitHub issues](https://github.com/prettier/prettier/issues) in order to find the best common rules for everyone**. (And it offers configuration options for the points about which the debates couldn't reach a conclusion.)

Prettier is considered by many to be the standard of web development (10M projects using it).

So maybe we could reverse your question: are you sure it's up to you to decide that your personal preferences are better for the people in your team than the standard developed by the web community?

(Edit: just a note, if you have set up your ESLint to not put orange squiggles everywhere and to automatically format the code (essentially reproducing Prettier's behaviour), then that's pretty good. The problem, in many places refusing Prettier, is that automation is very bad, and PRs get dozens of useless comments about formatting.)

**Reply by alexleduc76** [reply]
@Zwyx  Well, now that ESLint has moved the formatting rules to a separate repo, we can know how much using custom formatting rules matter to other developers (VS using ESLint strictly as a linter).

According to NPM, @stylistic/eslint-plugin is downloaded 1.4 million times per week. This means there are 1.4 million new projects per week that explicitly dislike the standard imposed by opinionated formatters, since there is no other reason to use that package. It doesn't look like code formatting is a settled debate at all.

**Reply by Zwyx** [reply]
Thanks Alex. I see your point, however Prettier is 52M downloads per week. So 97% Prettier, 3% ESLint Stylistic. I'm not sure at which point we can consider a debate settled, but at least we can say there is a clear preference.

(And that doesn't count the projects using Prettier but never downloading it from NPM, as the VS Code extension comes bundled with it. Personally I don't bother installing it for tiny projects, so I'd say half of what I do uses Prettier without downloading it.)

**Comment by Boshen** [comment, marked as answer]
Two years have passed since the previous reply.

Our formatter is now ready https://oxc.rs/docs/guide/usage/formatter.html

And I believe Oxc have excelled in most areas today.
